### ISSUE TITLE
[RFC] Literate Haskell semantics

### ISSUE THREAD

**Issue by Nadrieril** [dhall-lang/dhall-lang#959]
### TL;DR

I propose we write the inference rules of the standard in literate Haskell, so that they become executable. That would make experimentation easier and the standard more robust.

### Proposal

The way we currently write the standard is not the most convenient or robust. Among several issues, inference rules aren't very natural to read or write and are difficult to abstract over. Some bits of the standard (notably the imports spec and the typing rule for merge) are notoriously hard to understand. We rely on implementations to validate tests and ensure the standard is consistent.

I propose we instead write the standard in literate Haskell. I believe most contributors here are at least somewhat familiar with Haskell, and in any case we shall not make use of advanced features, to ensure that the standard stays readable to non-Haskellers. The text of the standard need not change, but the infernce rules would be replaced.

For example, the typing rule

    Γ₀ ⊢ A : i   ↑(1, x, 0, (Γ₀, x : A)) = Γ₁   Γ₁ ⊢ B : o   i ↝ o : c
    ──────────────────────────────────────────────────────────────────
    Γ₀ ⊢ ∀(x : A) → B : c

could become

```haskell
typeOf ctx (Pi x a b) = do
	i <- typeOf ctx a
	let ctx1 = push x a ctx
	o <- typeOf ctx1 b
	c <- function_check i o
	return c
```

Benefits:
- easier to experiment with new features
- more readable semantics
- force us to be more precise, eg. around parsing and desugaring
- easier to translate the standard into code (since it is already code)
- tests are guaranteed to be consistent with the spec

Drawbacks:
- larger maintenance burden for contributors

This is a large undertaking and will take some time and effort. I hope we can reuse a lot of code from `dhall-haskell`, to avoid reinventing everything. We can get there gradually and the benefits will hopefully be worth it.

**Comment by Gabriel439** [comment]
I support doing this!  I'm also available to help with this if desired

I think it will be far easier to directly translate the judgments to Haskell code than to reuse the Haskell implementation.  The Haskell implementation is cluttered by a whole bunch of things that are not related to the spec (like customizability and formatting-related concerns).

**Comment by Nadrieril** [comment]
I see a lot of enthusiasm !

@Gabriel439 May I ask you to do the basic setup of a Haskell project ? It's been a while since I've written any Haskell from scratch.

The other bottleneck for having something useful is the parser. Obviously the parser should be derived from the abnf grammar. However [the only Haskell library to parse abnf](https://hackage.haskell.org/package/abnf) is quite unmaintained and depends on megaparsec 5.3.1 which is super old. I'm not sure what to do, I was thinking we might generate the CST in python using https://github.com/erikrose/parsimonious, output it in a simple format like s-exprs and then do the rest in Haskell.

**Comment by philandstuff** [comment]
Given this issue is about the *semantics*, I think having a hand-rolled parser that attempts to match the abnf is perfectly fine for now. This would allow us to break the problem down into two smaller chunks.

**Comment by Nadrieril** [comment]
Well, my intentions are definitely about having as much as possible of the standard made executable. Hand-rolling and maintaining a parser is a huge effort, I'd rather avoid that :/ But yeah if we have no other way that's an option

**Comment by Gabriel439** [comment]
@Nadrieril: Yeah, I can help set up the project skeleton to get things going

**Comment by Gabriel439** [comment]
@Nadrieril: Alright, I have a skeleton project ready for you:

https://github.com/dhall-lang/dhall-lang/compare/gabriel/reference

I set it up so that you should be able to obtain everything you need (except `cabal`) using `nix-shell`.  The command I use for type-checking the project is:

```
$ cabal2nix --shell . > shell.nix && nix-shell --run 'ghcid --command "cabal repl"'
```

Basically, add the Literate Haskell code blocks directly to the markdown files, and then create `*.lhs` files which are symlinks to the markdown files and list them in the `standard.cabal` file.

I also included an example of using the `transform` utility from the `lens` package to automate the boilerplate of "just recurse over child sub-expressions" that is common in several sections of the standard:

https://github.com/dhall-lang/dhall-lang/compare/gabriel/reference#diff-1c7c5fb00174eebe31fce210f18be977R594-R596

I can pin things down further if you run into any issues with the `nix-shell`

**Comment by Nadrieril** [comment]
Hey, sorry for letting this slip away. Life happened and I've sadly been cutting down on time invested in Dhall. This issue in particular is a huge undertaking that I don't see myself having time for anytime soon :(. Anyone feel free to pick it up if you're excited about it!

**Comment by Gabriel439** [comment]
@Nadrieril: No worries!  I can pick this up

**Comment by Gabriel439** [comment]
I have the first PR up for this here: https://github.com/dhall-lang/dhall-lang/pull/1102

**Comment by Gabriel439** [comment]
I have the second PR up for review here: https://github.com/dhall-lang/dhall-lang/pull/1127

**Comment by Gabriel439** [comment]
I forgot to link this other PR which is relevant to this issue: https://github.com/dhall-lang/dhall-lang/pull/1152

**Comment by darichey** [comment]
Hi, I have two questions regarding this effort:
1. Is the lack of Haskell code for any given part of the standard simply due to it not being done yet? If so, would PRs be welcome? For example, for type-inference.md.
2. Is it a goal to run the acceptance tests suite against the reference implementation? I see https://github.com/dhall-lang/dhall-lang/blob/35c115c392a9101096d20ec289d12394d1b4b5d9/standard/tasty/Main.hs, but I _think_ it's not testing everything that has a reference impl.

**Comment by Gabriel439** [comment]
Yeah, any missing Haskell code is just due to it not being done yet.  I got started and then got distracted and forgot to finish things

Also, yes, the goal is to run the acceptance tests against the reference implementation

**Comment by winitzki** [comment]
I think we need to have a fully standard-compliant reference implementation in literate Haskell. Performance doesn't have to be good. This could be a great starting point for all language implementations, tests, and learning about Dhall.

**Comment by Gabriella439** [comment]
Note that there already is the beginnings of that in the `./standard` directory: https://github.com/dhall-lang/dhall-lang/tree/master/standard.  It's not yet the complete standard, but it is written in literate Haskell

**Comment by winitzki** [comment]
Yes, I know, and that's what I was thinking about. So far I've made a number of prompts for AI towards creating a complete reference implementation. https://github.com/dhall-lang/dhall-lang/pull/1440  Not sure if this is a good approach.
